Jenkinsfile

```diff
@@ -0,0 +1,335 @@
+  agent any
+  environment {
+    HOME = "${env.HUDSON_HOME}"
+    BASE_DIR="src/go.elastic.co/apm"
+    JOB_GIT_CREDENTIALS = "f6c7695a-671e-4f4f-a331-acdce44ff9ba"
+  }
+  triggers {
+    cron('0 0 * * 1-5')
+    githubPush()
+  }
+  options {
+    timeout(time: 1, unit: 'HOURS') 
+    buildDiscarder(logRotator(numToKeepStr: '3', artifactNumToKeepStr: '2', daysToKeepStr: '30'))
+    timestamps()
+    preserveStashes()
+    ansiColor('xterm')
+    disableResume()
+    durabilityHint('PERFORMANCE_OPTIMIZED')
+  }
+  parameters {
+    string(name: 'branch_specifier', defaultValue: "", description: "the Git branch specifier to build (<branchName>, <tagName>, <commitId>, etc.)")    
+    booleanParam(name: 'linux_ci', defaultValue: true, description: 'Enable Linux build')
+    booleanParam(name: 'test_ci', defaultValue: true, description: 'Enable test')
+    booleanParam(name: 'integration_test_ci', defaultValue: true, description: 'Enable run integgration test')
+    booleanParam(name: 'integration_test_pr_ci', defaultValue: true, description: 'Enable run integgration test')
+    booleanParam(name: 'integration_test_master_ci', defaultValue: true, description: 'Enable run integgration test')
+    booleanParam(name: 'bench_ci', defaultValue: true, description: 'Enable benchmarks')
+    booleanParam(name: 'doc_ci', defaultValue: true, description: 'Enable build documentation')
+  }
+  
+  stages {
+    /**
+     Checkout the code and stash it, to use it on other stages.
+    */
+    stage('Checkout') {
+      agent { label 'master || linux' }
+      environment {
+        PATH = "${env.PATH}:${env.HUDSON_HOME}/go/bin/:${env.WORKSPACE}/bin"
+        GOPATH = "${env.WORKSPACE}"
+      }
+      
+      steps {
+        withEnvWrapper() {
+          dir("${BASE_DIR}"){
+            script{
+              if(!branch_specifier){
+                echo "Checkout SCM ${GIT_BRANCH}"
+                checkout scm
+              } else {
+                echo "Checkout ${branch_specifier}"
+                checkout([$class: 'GitSCM', branches: [[name: "${branch_specifier}"]], 
+                  doGenerateSubmoduleConfigurations: false, 
+                  extensions: [],
+                  submoduleCfg: [],
+                  userRemoteConfigs: [[credentialsId: "${JOB_GIT_CREDENTIALS}", 
+                  url: "${GIT_URL}"]]])
+              }
+              env.JOB_GIT_COMMIT = getGitCommitSha()
+              env.JOB_GIT_URL = "${GIT_URL}"
+              
+              github_enterprise_constructor()
+              
+              on_change{
+                echo "build cause a change (commit or PR)"
+              }
+              
+              on_commit {
+                echo "build cause a commit"
+              }
+              
+              on_merge {
+                echo "build cause a merge"
+              }
+              
+              on_pull_request {
+                echo "build cause PR"
+              }
+              gitCreateTag()
```

Comment:
What's the tag going to be? I'd generally prefer that CI does not modify the repo.

Reply:
it creates a tag to use in the integration test to test against the exact version, it is deleted at the end of the job. It is easy to work with tags instead commit sha1, and all the integration test that we have right now are based on branchs, so you can use also tags.

Reply:
So this just gets passed down to apm-integration-testing, to specify the branch? If I update apm-integration-testing to accept a commit instead of (or as well as) a branch, can we get rid of the tagging?

The branch is currently ignored by the Go service in apm-integration-testing anyway, so I need to make updates either way.

Reply:
I made the changes to manage branches, as do all the modules, we can open a team discussion about use tags or commits, but it only push tags so does not change anything else I do not see it as a problem, tags are for this kind of thing.

Reply:
Let's bring it up with the team.

Go's tooling uses tags for identifying module versions, and GitHub uses tags for releases; using them for transient state feels odd to me if/when we can propagate the commit. Also, perhaps I'm a bit overcautious, but I like CI to have as few permissions as possible. Being able to write, even if it does not write, makes me nervous.

Reply:
Sure 😄